esp/services/esdl_svc_engine/esdl_monitor.cpp

```diff
@@ -0,0 +1,576 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "esdl_monitor.hpp"
+#include "jmd5.hpp"
+#include "esdl_binding.hpp"
+#include "esdl_svc_engine.hpp"
+#include <memory>
+
+class CEsdlInstance : public CInterface
+{
+public:
+    CEsdlInstance() { }
+    CEsdlInstance(const char* defId, IEsdlDefinition* def)
+    {
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        m_defId.set(lcid.str());
+        m_def.set(def);
+    }
+    virtual ~CEsdlInstance()
+    {
+    }
+    StringBuffer m_defId;
+    Owned<IEsdlDefinition> m_def;
+};
+
+//This class is for storing esdl definition objects to be shared among
+//ESDL services to save memory space and increase efficiency
+class CEsdlShare : implements IEsdlShare, public Thread
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    CEsdlShare() : m_stop(false)
+    {
+    }
+    virtual ~CEsdlShare() {}
+
+    virtual void add(const char* defId, IEsdlDefinition* def) override
+    {
+        if (!defId || !*defId || !def)
+            return;
+        DBGLOG("Adding esdl definition %s to shared cache", defId);
+        Owned<CEsdlInstance> instance = new CEsdlInstance(defId, def);
+        CriticalBlock cb(m_critSect);
+        m_esdlInstanceMap.setValue(defId, instance.getLink());
+    }
+
+    virtual void remove(const char* defId) override
+    {
+        if (!defId || !*defId)
+            return;
+        StringBuffer lcid(defId);
+        lcid.trim().toLowerCase();
+        if (lcid.length() == 0)
+            return;
+        defId = lcid.str();
+        DBGLOG("Removing esdl definition %s from shared cache", defId);
+        {
+            CriticalBlock cb(m_critSect);
+            m_esdlInstanceMap.remove(defId);
+        }
+    }
+
+    virtual Linked<IEsdlDefinition> query(const char* defId) override
```

Comment:
this looks a bit questionable, especially given method name is query.
Shouldn't it just be:
virtual IEsdlDefinition *query(const char* defId) override

Reply:
The purpose of returning Linked is to hold on to the object right away, because there's a background thread that could clean up the object at any instant. I'll change the "query" method name which is too suggestive to a more neutral name "lookup".

Reply:
The purpose for returning Linked is to hold on to the definition right away, because there's a cleaning up thread that may delete the definition at any instant. I agree that the "query" method name is suggestive, so I'll change it to more neutral "lookup".